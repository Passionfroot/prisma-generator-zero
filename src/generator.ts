import { GeneratorConfig, generatorHandler, GeneratorOptions } from "@prisma/generator-helper";
import { writeFile, mkdir } from "fs/promises";
import { join } from "path";
import { version } from "../package.json";
import { Config } from "./types";
import { transformSchema } from "./mappers/schemaMapper";
import { generateCode } from "./generators/codeGenerator";

export async function onGenerate(options: GeneratorOptions) {
  const { generator, dmmf } = options;
  const outputFile = "schema.ts";
  const outputDir = generator.output?.value;

  if (!outputDir) {
    throw new Error("Output directory is required");
  }

  const config = {
    name: generator.name,
    prettier: parseBooleanConfig(generator.config.prettier, false),
    resolvePrettierConfig: parseBooleanConfigDefaultTrue(generator.config.resolvePrettierConfig),
    remapTablesToCamelCase: parseBooleanConfig(generator.config.remapTablesToCamelCase, false),
    remapColumnsToCamelCase: parseBooleanConfig(generator.config.remapColumnsToCamelCase, false),
    excludeTables: loadExcludeTables(generator),
    enumAsUnion: parseBooleanConfig(generator.config.enumAsUnion, false),
  } satisfies Config;

  const transformedSchema = transformSchema(dmmf, config);

  let output = generateCode(transformedSchema, config);

  if (config.prettier) {
    let prettier: typeof import("prettier");
    try {
      prettier = await import("prettier");
    } catch {
      throw new Error("Unable import Prettier. Is it installed?");
    }

    const prettierOptions = config.resolvePrettierConfig
      ? await prettier.resolveConfig(outputFile)
      : null;

    output = await prettier.format(output, { ...prettierOptions, parser: "typescript" });
  }

  await mkdir(outputDir, { recursive: true });
  await writeFile(join(outputDir, outputFile), output);
}

generatorHandler({
  onManifest() {
    return {
      version,
      defaultOutput: "generated/zero",
      prettyName: "Zero Schema",
    };
  },
  onGenerate,
});

/**
 * Load the excludeTables from the generator config
 */
function loadExcludeTables(generator: GeneratorConfig) {
  const value = generator.config.excludeTables;

  if (value === undefined) {
    return [];
  }

  if (!Array.isArray(value)) {
    throw new Error("excludeTables must be an array");
  }

  return value;
}
/**
 * Parses a boolean config value from Prisma generator config where the default is false.
 * Handles boolean true, string "true", or returns false otherwise.
 */
function parseBooleanConfig(value: any, defaultValue: boolean): boolean {
  if (value === undefined) {
    return defaultValue;
  }
  return value === true || value === "true";
}

/**
 * Parses a boolean config value where the default is true.
 * Handles boolean false, string "false", or returns true otherwise.
 */
function parseBooleanConfigDefaultTrue(value: any): boolean {
  if (value === undefined) {
    return true;
  }
  return !(value === false || value === "false");
}
