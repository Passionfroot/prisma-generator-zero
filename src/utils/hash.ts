import type { DMMF } from "@prisma/generator-helper";
import { createHash } from "crypto";

export function generateSchemaHash(models: DMMF.Model[], enums: DMMF.DatamodelEnum[]): string {
  const hash = createHash("sha256");

  // Only hash the structural elements that affect the schema
  const schemaStructure = {
    models: models.map((model) => ({
      name: model.name,
      dbName: model.dbName,
      fields: model.fields.map((f) => ({
        name: f.name,
        type: f.type,
        isRequired: f.isRequired,
        isList: f.isList,
        relationName: f.relationName,
        relationFromFields: f.relationFromFields,
        relationToFields: f.relationToFields,
        default: f.default,
        unique: f.isUnique,
      })),
      primaryKey: model.primaryKey,
      uniqueFields: model.uniqueFields,
      uniqueIndexes: model.uniqueIndexes,
    })),
    enums: enums.map((enumType) => ({
      name: enumType.name,
      values: enumType.values.map((v) => ({
        name: v.name,
        dbName: v.dbName,
      })),
    })),
  };

  hash.update(JSON.stringify(schemaStructure));
  return hash.digest("hex");
} 
