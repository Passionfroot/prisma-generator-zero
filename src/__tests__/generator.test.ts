import { describe, it, expect, beforeEach, vi } from "vitest";
import { DMMF, GeneratorOptions } from "@prisma/generator-helper";
import * as fs from "fs/promises";
import { onGenerate } from "../generator";
import { createField, createModel, createMockDMMF } from "./utils";

// Mock fs/promises
vi.mock("fs/promises", () => ({
  writeFile: vi.fn(),
  mkdir: vi.fn(),
  readFile: vi.fn(),
}));

function createTestOptions(dmmf: DMMF.Document): GeneratorOptions {
  return {
    generator: {
      output: { value: "generated", fromEnvVar: null },
      name: "test-generator",
      config: {},
      provider: { value: "test-provider", fromEnvVar: null },
      binaryTargets: [],
      previewFeatures: [],
      sourceFilePath: "",
    },
    dmmf,
    schemaPath: "",
    datasources: [],
    otherGenerators: [],
    version: "0.0.0",
    datamodel: "",
  };
}

describe("Generator", () => {
  beforeEach(() => {
    vi.clearAllMocks();
  });

  describe("Schema Generation", () => {
    it("should generate correct schema for basic model", async () => {
      const mockModel: DMMF.Model = {
        name: "User",
        dbName: null,
        fields: [
          createField("id", "String", { isId: true }),
          createField("name", "String"),
          createField("email", "String"),
          createField("age", "Int", { isRequired: false }),
        ],
        uniqueFields: [],
        uniqueIndexes: [],
        primaryKey: null,
      };

      await onGenerate(createTestOptions(createMockDMMF([mockModel])));

      expect(fs.mkdir).toHaveBeenCalledWith("generated", { recursive: true });

      const writeFileCalls = vi.mocked(fs.writeFile).mock.calls;
      expect(writeFileCalls.length).toBe(1);

      const [, contentBuffer] = writeFileCalls[0];
      const content = contentBuffer.toString();

      expect(content).toMatchSnapshot();
    });

    it("should handle enums correctly", async () => {
      const mockEnum: DMMF.DatamodelEnum = {
        name: "Role",
        values: [
          { name: "USER", dbName: null },
          { name: "ADMIN", dbName: null },
        ],
        dbName: null,
      };

      const mockModel: DMMF.Model = {
        name: "User",
        dbName: null,
        fields: [
          createField("id", "String", { isId: true }),
          createField("role", "Role", { kind: "enum" }),
        ],
        uniqueFields: [],
        uniqueIndexes: [],
        primaryKey: null,
      };

      await onGenerate(
        createTestOptions(createMockDMMF([mockModel], [mockEnum])),
      );

      const [, contentBuffer] = vi.mocked(fs.writeFile).mock.calls[0];
      const content = contentBuffer.toString();

      expect(content).toMatchSnapshot();
    });

    it("should handle enums as unions correctly", async () => {
      const mockEnum: DMMF.DatamodelEnum = {
        name: "Role",
        values: [
          { name: "USER", dbName: null },
          { name: "ADMIN", dbName: null },
        ],
        dbName: null,
      };

      const mockModel: DMMF.Model = {
        name: "User",
        dbName: null,
        fields: [
          createField("id", "String", { isId: true }),
          createField("role", "Role", { kind: "enum" }),
        ],
        uniqueFields: [],
        uniqueIndexes: [],
        primaryKey: null,
      };

      const options = createTestOptions(
        createMockDMMF([mockModel], [mockEnum]),
      );

      options.generator.config.enumAsUnion = "true";

      await onGenerate(options);

      const [, contentBuffer] = vi.mocked(fs.writeFile).mock.calls[0];
      const content = contentBuffer.toString();

      expect(content).toMatchSnapshot();
    });

    it("should handle relationships correctly", async () => {
      const userModel = createModel("User", [
        createField("id", "String", { isId: true }),
        createField("name", "String"),
        createField("posts", "Post", {
          kind: "object",
          isList: true,
          relationName: "UserPosts",
          relationToFields: ["id"],
          relationFromFields: ["userId"],
        }),
      ]);

      const postModel = createModel("Post", [
        createField("id", "String", { isId: true }),
        createField("title", "String"),
        createField("userId", "String"),
        createField("user", "User", {
          kind: "object",
          relationName: "UserPosts",
          relationFromFields: ["userId"],
          relationToFields: ["id"],
        }),
      ]);

      await onGenerate(createTestOptions(createMockDMMF([userModel, postModel])));

      const [, contentBuffer] = vi.mocked(fs.writeFile).mock.calls[0];
      const content = contentBuffer.toString();

      expect(content).toMatchSnapshot();
    });
  });

  it("should handle model mapping with @@map correctly", async () => {
    // model cdr @@map("xml_cdr")
    const mappedModel = createModel("cdr", [
      createField("id", "Int", { isId: true }),
      createField("data", "String"),
    ], { dbName: "xml_cdr" });

    const dmmf = createMockDMMF([mappedModel]);
    const options = createTestOptions(dmmf);
    // Assuming default remapTablesToCamelCase = false for this test

    await onGenerate(options);

    const writeFileCalls = vi.mocked(fs.writeFile).mock.calls;
    expect(writeFileCalls.length).toBe(1);
    const [, contentBuffer] = writeFileCalls[0];
    const content = contentBuffer.toString();

    expect(content).toContain('export const cdrTable = table("cdr")');
    expect(content).toContain('.from("xml_cdr")');
    expect(content).toContain('tables: [\n      cdrTable,\n    ],');
    expect(content).toContain('export type cdr = Row<typeof schema.tables.cdr>;');

    expect(content).toMatchSnapshot("model @@map test");
  });

  it("should handle column mapping with @map and remapColumnsToCamelCase correctly", async () => {
    const modelWithMappings = createModel("MappedProduct", [
      createField("id", "String", { isId: true }),
        createField("product_code", "String"),
        createField("priceAmount", "Float", { dbName: "price_in_db" }),
        createField("status", "String", { dbName: "current_status" }),
        createField("alreadyCamel", "String"),
      ]);

      const dmmf = createMockDMMF([modelWithMappings]);

      // --- Test Case 1: remapColumnsToCamelCase = false ---
      const optionsFalse = createTestOptions(dmmf);
      optionsFalse.generator.config.remapColumnsToCamelCase = "false";

      await onGenerate(optionsFalse);
      let writeFileCalls = vi.mocked(fs.writeFile).mock.calls;
      expect(writeFileCalls.length).toBe(1);
      let [, contentBufferFalse] = writeFileCalls[0];
      let contentFalse = contentBufferFalse.toString();

      expect(contentFalse).toContain('id: string()');
      expect(contentFalse).toContain('product_code: string()');
      expect(contentFalse).toContain('priceAmount: number().from("price_in_db")');
      expect(contentFalse).toContain('status: string().from("current_status")');
      expect(contentFalse).toContain('alreadyCamel: string()');

      vi.clearAllMocks();

      // --- Test Case 2: remapTablesToCamelCase = true (and columns) ---
      const optionsTrue = createTestOptions(dmmf);
      optionsTrue.generator.config.remapTablesToCamelCase = "true";
      optionsTrue.generator.config.remapColumnsToCamelCase = "true";

      await onGenerate(optionsTrue);
      writeFileCalls = vi.mocked(fs.writeFile).mock.calls;
      expect(writeFileCalls.length).toBe(1);
      let [, contentBufferTrue] = writeFileCalls[0];
      let contentTrue = contentBufferTrue.toString();

      expect(contentTrue).toContain('id: string()');
      expect(contentTrue).toContain('productCode: string().from("product_code")');
      expect(contentTrue).toContain('priceAmount: number().from("price_in_db")');
      expect(contentTrue).toContain('status: string().from("current_status")');
      expect(contentTrue).toContain('alreadyCamel: string()');

      expect(contentTrue).toContain('export type mappedProduct = Row<typeof schema.tables.mappedProduct>;');
    });

  describe("Many-to-Many Relationships", () => {
    it("should generate correct schema for implicit many-to-many relationship", async () => {
      const postModel = createModel("Post", [
        createField("id", "Int", { isId: true }),
        createField("title", "String"),
        createField("categories", "Category", {
          isList: true,

          kind: "object",
        }),
      ]);

      const categoryModel = createModel("Category", [
        createField("id", "Int", { isId: true }),
        createField("name", "String"),
        createField("posts", "Post", {
          isList: true,

          kind: "object",
        }),
      ]);

      const dmmf = createMockDMMF([postModel, categoryModel]);
      const options = createTestOptions(dmmf);

      // Mock readFile to return null (no existing schema)
      vi.mocked(fs.readFile).mockRejectedValue(new Error("File not found"));

      await onGenerate(options);

      const writeFileCall = vi.mocked(fs.writeFile).mock.calls[0];
      const generatedCode = writeFileCall[1] as string;

      expect(generatedCode).toMatchSnapshot();
    });

    it("should use custom relation name for implicit many-to-many table", async () => {
      const postModel = createModel("Post", [
        createField("id", "Int", { isId: true }),
        createField("title", "String"),
        createField("categories", "Category", {
          isList: true,
          relationName: "MyCustomRelation",
          kind: "object",
        }),
      ]);

      const categoryModel = createModel("Category", [
        createField("id", "Int", { isId: true }),
        createField("name", "String"),
        createField("posts", "Post", {
          isList: true,
          relationName: "MyCustomRelation",
          kind: "object",
        }),
      ]);

      const dmmf = createMockDMMF([postModel, categoryModel]);
      const options = createTestOptions(dmmf);

      // Mock readFile to return null (no existing schema)
      vi.mocked(fs.readFile).mockRejectedValue(new Error("File not found"));

      await onGenerate(options);

      const writeFileCall = vi.mocked(fs.writeFile).mock.calls[0];
      const generatedCode = writeFileCall[1] as string;

      expect(generatedCode).toMatchSnapshot();
    });
  });
});
