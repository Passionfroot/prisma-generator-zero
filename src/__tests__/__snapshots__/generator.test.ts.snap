// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Generator > Many-to-Many Relationships > should generate correct schema for implicit many-to-many relationship 1`] = `
"
import {
  table,
  string,
  boolean,
  number,
  json,
  enumeration,
  relationships,
  createSchema,
  type Row,
} from "@rocicorp/zero";


export const PostTable = table("Post")
  .columns({
    id: number(),
    title: string(),
  })
  .primaryKey("id");

export const CategoryTable = table("Category")
  .columns({
    id: number(),
    name: string(),
  })
  .primaryKey("id");


export const schema = createSchema(
  {
    tables: [
      PostTable,
      CategoryTable,
    ],
  }
);

export type Schema = typeof schema;
export type Post = Row<typeof schema.tables.Post>;
export type Category = Row<typeof schema.tables.Category>;
"
`;

exports[`Generator > Many-to-Many Relationships > should use custom relation name for implicit many-to-many table 1`] = `
"
import {
  table,
  string,
  boolean,
  number,
  json,
  enumeration,
  relationships,
  createSchema,
  type Row,
} from "@rocicorp/zero";


export const PostTable = table("Post")
  .columns({
    id: number(),
    title: string(),
  })
  .primaryKey("id");

export const CategoryTable = table("Category")
  .columns({
    id: number(),
    name: string(),
  })
  .primaryKey("id");

export const _MyCustomRelationTable = table("_MyCustomRelation")
  .columns({
    A: string(),
    B: string(),
  })
  .primaryKey("A", "B");


export const PostTableRelationships = relationships(PostTable, ({ many }) => ({
  categories: many({
    sourceField: ["id"],
    destField: ["B"],
    destSchema: _MyCustomRelationTable,
  }, {
    sourceField: ["A"],
    destField: ["id"],
    destSchema: CategoryTable,
  })
}));

export const CategoryTableRelationships = relationships(CategoryTable, ({ many }) => ({
  posts: many({
    sourceField: ["id"],
    destField: ["A"],
    destSchema: _MyCustomRelationTable,
  }, {
    sourceField: ["B"],
    destField: ["id"],
    destSchema: PostTable,
  })
}));

export const _MyCustomRelationTableRelationships = relationships(_MyCustomRelationTable, ({ one }) => ({
  modelA: one({
    sourceField: ["A"],
    destField: ["id"],
    destSchema: categoryTable,
  }),
  modelB: one({
    sourceField: ["B"],
    destField: ["id"],
    destSchema: postTable,
  })
}));


export const schema = createSchema(
  {
    tables: [
      PostTable,
      CategoryTable,
      _MyCustomRelationTable,
    ],
    relationships: [
      PostTableRelationships,
      CategoryTableRelationships,
      _MyCustomRelationTableRelationships,
    ],
  }
);

export type Schema = typeof schema;
export type Post = Row<typeof schema.tables.Post>;
export type Category = Row<typeof schema.tables.Category>;
export type _MyCustomRelation = Row<typeof schema.tables._MyCustomRelation>;
"
`;

exports[`Generator > Schema Generation > should generate correct schema for basic model 1`] = `
"
import {
  table,
  string,
  boolean,
  number,
  json,
  enumeration,
  relationships,
  createSchema,
  type Row,
} from "@rocicorp/zero";


export const UserTable = table("User")
  .columns({
    id: string(),
    name: string(),
    email: string(),
    age: number().optional(),
  })
  .primaryKey("id");


export const schema = createSchema(
  {
    tables: [
      UserTable,
    ],
  }
);

export type Schema = typeof schema;
export type User = Row<typeof schema.tables.User>;
"
`;

exports[`Generator > Schema Generation > should handle enums as unions correctly 1`] = `
"
import {
  table,
  string,
  boolean,
  number,
  json,
  enumeration,
  relationships,
  createSchema,
  type Row,
} from "@rocicorp/zero";


export type Role = "USER" | "ADMIN";


export const UserTable = table("User")
  .columns({
    id: string(),
    role: enumeration<Role>(),
  })
  .primaryKey("id");


export const schema = createSchema(
  {
    tables: [
      UserTable,
    ],
  }
);

export type Schema = typeof schema;
export type User = Row<typeof schema.tables.User>;
"
`;

exports[`Generator > Schema Generation > should handle enums correctly 1`] = `
"
import {
  table,
  string,
  boolean,
  number,
  json,
  enumeration,
  relationships,
  createSchema,
  type Row,
} from "@rocicorp/zero";


export enum Role {
  USER = "USER",
  ADMIN = "ADMIN",
}


export const UserTable = table("User")
  .columns({
    id: string(),
    role: enumeration<Role>(),
  })
  .primaryKey("id");


export const schema = createSchema(
  {
    tables: [
      UserTable,
    ],
  }
);

export type Schema = typeof schema;
export type User = Row<typeof schema.tables.User>;
"
`;

exports[`Generator > Schema Generation > should handle relationships correctly 1`] = `
"
import {
  table,
  string,
  boolean,
  number,
  json,
  enumeration,
  relationships,
  createSchema,
  type Row,
} from "@rocicorp/zero";


export const UserTable = table("User")
  .columns({
    id: string(),
    name: string(),
  })
  .primaryKey("id");

export const PostTable = table("Post")
  .columns({
    id: string(),
    title: string(),
    userId: string(),
  })
  .primaryKey("id");


export const UserTableRelationships = relationships(UserTable, ({ many }) => ({
  posts: many({
    sourceField: ["id"],
    destField: ["userId"],
    destSchema: PostTable,
  })
}));

export const PostTableRelationships = relationships(PostTable, ({ one }) => ({
  user: one({
    sourceField: ["userId"],
    destField: ["id"],
    destSchema: UserTable,
  })
}));


export const schema = createSchema(
  {
    tables: [
      UserTable,
      PostTable,
    ],
    relationships: [
      UserTableRelationships,
      PostTableRelationships,
    ],
  }
);

export type Schema = typeof schema;
export type User = Row<typeof schema.tables.User>;
export type Post = Row<typeof schema.tables.Post>;
"
`;

exports[`Generator > should handle model mapping with @@map correctly > model @@map test 1`] = `
"
import {
  table,
  string,
  boolean,
  number,
  json,
  enumeration,
  relationships,
  createSchema,
  type Row,
} from "@rocicorp/zero";


export const cdrTable = table("cdr")
  .from("xml_cdr")
  .columns({
    id: number(),
    data: string(),
  })
  .primaryKey("id");


export const schema = createSchema(
  {
    tables: [
      cdrTable,
    ],
  }
);

export type Schema = typeof schema;
export type cdr = Row<typeof schema.tables.cdr>;
"
`;
